docs/src/index.md

```diff
@@ -4,6 +4,350 @@ DocTestSetup = :(using ChainRulesCore, ChainRules)
+
+ - `Thunk`: this is a deferred computation. A thunk is a [word for a zero argument closure](https://en.wikipedia.org/wiki/Thunk). A computation wrapped in a `@thunk` doesn't get evaluated until `extern` is called on the `Thunk`. More on thunks later.
+ - `One`, `Zero`: There are special representations of `1` and `0`. They do great things around avoiding expanding `Thunks` in multiplication and (for `Zero`) addition.
+
+#### Other `AbstractDifferential`s: don't worry about them right now
+ - `Wirtinger`: it is complex. The docs need to be better. [Read the links in this issue](https://github.com/JuliaDiff/ChainRulesCore.jl/issues/40).
+ - `Casted`: it implements broadcasting mechanics. See [#10](https://github.com/JuliaDiff/ChainRulesCore.jl/issues/10)
+ - `InplaceableThunk`: it is like a Thunk but it can do `store!` and `accumulate!` in-place.
+
+ -------------------------------
+
+## Example of using ChainRules directly.
+
+While ChainRules is largely intended as a backend for autodiff systems, it can be used directly.
+In fact, this can be very useful if you can constrain the code you need to differentiate to only use things that have rules defined for.
+This was once how all neural network code worked.
+
+Using ChainRules directly also helps get a feel for it.
+
+```julia
+using ChainRules
+
+function foo(x)
+    a = sin(x)
+    b = 2a
+    c = asin(b)
+    return c
+end
+
+#### Find dfoo/dx via rrules
+
+# First the forward pass, accumulating rules
+x = 3;
+a, a_pullback = rrule(sin, x);
+b, b_pullback = rrule(*, 2, a);
+c, c_pullback = rrule(asin, b)
+
+# Then the backward pass calculating gradients
+c̄ = 1;  # ∂c/∂c
+_, b̄ = c_pullback(extern(c̄));     # ∂c/∂b
+_, _, ā = b_pullback(extern(b̄));  # ∂c/∂a
+_, x̄ = a_pullback(extern(ā));     # ∂c/∂x = ∂f/∂x
+extern(x̄)
+# -2.0638950738662625
+
+#### Find dfoo/dx via frules
+
+# Unlike with rrule, we can interleave evaluation and derivative evaluation
+x = 3;
+ẋ = 1;  # ∂x/∂x
+nofields = NamedTuple();  # ∂self/∂self
+
+a, a_pushforward = frule(sin, x);
+ȧ = a_pushforward(nofields, extern(ẋ));     # ∂a/∂x
+
+b, b_pushforward = frule(*, 2, a);
+ḃ = b_pushforward(nofields, 0, extern(ȧ));  # ∂b/∂x = ∂b/∂a⋅∂a/∂x
+
+c, c_pushforward = frule(asin, b);
+ċ = c_pushforward(nofields, extern(ḃ));     # ∂c/∂x = ∂c/∂b⋅∂b/∂x = ∂f/∂x
+extern(ċ)
+# -2.0638950738662625
+
+#### Find dfoo/dx via finite-differences
+
+using FiniteDifferences
+central_fdm(5, 1)(foo, x)
+# -2.0638950738670734
+
+#### Find dfoo/dx via finite-differences ForwardDiff.jl
+using ForwardDiff
+ForwardDiff.derivative(foo, x)
+# -2.0638950738662625
+
+#### Find dfoo/dx via finite-differences Zygote.jl
+using Zygote
+Zygote.gradient(foo, x)
+# (-2.0638950738662625,)
+```
+
+ -------------------------------
+
+## On writing good `rrule` / `frule` methods
+
+### Return Zero or One
+
+Rather than `0` or `1` or even rather than `zeros(n)`, `ones(n)`, or the identity matrix `I`.
```

Comment:
A verb in this sentence would be helpful. I guess it should be read as one sentence together with the header but it looks weird.

Reply:
Yeah, seems like half this sentence is missing.

Reply:
```suggestion
The `Zero()` and `One()` differential objects exist as an alternative to directly returning
`0` or `zeros(n)`, and `1` or `I`.
They allow more optimal computation when chaining pullbacks/pushforwards, to avoid work.
They should be used where possible.
```